benchmarks/src/test/java/org/apache/druid/benchmark/ProtobufParserBenchmark.java

```diff
@@ -0,0 +1,173 @@
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.concurrent.TimeUnit;
+
+@State(Scope.Benchmark)
+@Fork(value = 1)
+@Warmup(iterations = 10)
+@Measurement(iterations = 25)
+public class ProtobufParserBenchmark
+{
+  @Param({"75000"})
+  private int rowsPerSegment;
+
+  private static final Logger log = new Logger(ProtobufParserBenchmark.class);
+
+  static {
+    NullHandling.initializeForTests();
+  }
+
+  private ParseSpec nestedParseSpec;
+  private ProtobufInputRowParser nestedParser;
+  private ParseSpec flattenParseSpec;
+  private ProtobufInputRowParser flattenParser;
+  private byte[] protoInputs;
+  private String protoFilePath;
+
+  @Setup
+  public void setup()
+  {
+    log.info("SETUP CALLED AT " + +System.currentTimeMillis());
+
+    nestedParseSpec = new JSONParseSpec(
+                new TimestampSpec("timestamp", "iso", null),
+                new DimensionsSpec(Lists.newArrayList(
+                        new StringDimensionSchema("event"),
+                        new StringDimensionSchema("id"),
+                        new StringDimensionSchema("someOtherId"),
+                        new StringDimensionSchema("isValid")
+                ), null, null),
+                new JSONPathSpec(
+                        true,
+                        Lists.newArrayList(
+                                new JSONPathFieldSpec(JSONPathFieldType.ROOT, "eventType", "eventType"),
+                                new JSONPathFieldSpec(JSONPathFieldType.PATH, "foobar", "$.foo.bar"),
+                                new JSONPathFieldSpec(JSONPathFieldType.PATH, "bar0", "$.bar[0].bar")
+                        )
+                ),
+                null,
+                null
+    );
+
+    flattenParseSpec = new JSONParseSpec(
+            new TimestampSpec("timestamp", "iso", null),
+            new DimensionsSpec(Lists.newArrayList(
+                    new StringDimensionSchema("event"),
+                    new StringDimensionSchema("id"),
+                    new StringDimensionSchema("someOtherId"),
+                    new StringDimensionSchema("isValid")
+            ), null, null),
+            null,
+            null,
+            null
+    );
+
+    protoFilePath = "ProtoFile";
+    protoInputs = getProtoInputs(protoFilePath);
+    nestedParser = new ProtobufInputRowParser(nestedParseSpec, "prototest.desc", "ProtoTestEvent");
+    flattenParser = new ProtobufInputRowParser(flattenParseSpec, "prototest.desc", "ProtoTestEvent");
+  }
+
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.MICROSECONDS)
+  public void consumeFlattenData(Blackhole blackhole)
+  {
+    for (int i = 0; i < rowsPerSegment; i++) {
+      InputRow row = flattenParser.parseBatch(ByteBuffer.wrap(protoInputs)).get(0);
```

Comment:
I wonder what is the effect of `ByteBuffer.wrap()` on the measured performance.
Is there a reason the wrapping is done inside the loop and not in the setup phase?

Reply:
Yeah your words make sense. I would put the wrapping function in the setup phase. It would make the benchmark easier to understand and be clearer.

Reply:
I see. It makes sense. It could be solved by `buff.pos(0)`, but since `ByteBuffer.wrap()` only instantiate a single object with O(1) complexity, I don't think it would make much of a difference.